### PR TITLE
feat(organizations): scope assignments under organization group URLs

### LINKED ISSUE #7721: Feature: Scope assignments under organization-owned group URLs
### Is your feature request related to a problem? Please describe

Organization-owned groups can now be accessed through organization-scoped URLs **(/organizations/:org/groups/:id)**, but their assignments still only work through the flat, non-org-scoped routes (/groups/:group_id/assignments/...). This is inconsistent: when a mentor or admin is working inside an organization context and opens a group's assignments, the URLs drop the organization scope. It also means organization admins and mentors cannot manage assignments in classrooms created by other mentors, because assignment authorization only recognizes the group's primary mentor, not organization-level roles.

### Describe the solution you'd like

Make assignments work fully under organization-scoped group URLs, matching how the groups themselves already behave:

- Assignments in organization-owned groups should be accessible at /organizations/:org/groups/:group_id/assignments/... (view, create, edit, delete, reopen, close, start).
- Standalone groups (not part of any organization) should continue to use their existing flat assignment URLs, unchanged.
- Organization admins and assigned mentors should be able to manage assignments in any group within their organization, not just groups where they are the primary mentor.
- After creating, editing, or deleting an assignment in an organization group, the user should be redirected back to the organization-scoped group page.
- Assignment links throughout the group and assignment views should point to the correct scoped URL based on whether the group belongs to an organization.

### Describe alternatives you've considered

N/A

### Additional context

Project 6 Enterprise & Institutional Organization Features

### Are you working on this?

Yes

### PR DESCRIPTION
**Pull request by naman79820**
Fixes #7721

 

#### Describe the changes you have made in this PR -
This PR makes assignments work under organization-scoped group URLs, matching how organization-owned groups already behave.

- Assignments in organization-owned groups are now accessible at /organizations/:org/groups/:group_id/assignments/... for all actions (view, create, edit, delete, reopen, close, start). Standalone groups keep their existing flat assignment URLs unchanged.
- Organization admins and assigned mentors can now manage assignments in any group within their organization, not only groups where they are the primary mentor.
- After creating, editing, or deleting an assignment in an organization group, the user is redirected back to the organization-scoped group page.
- Assignment links across the group and assignment views now resolve to the correct scoped URL depending on whether the group belongs to an organization, using shared helpers.
- Tenant isolation: assignment and group lookups are scoped through the organization in the URL, so a request can't reach another organization's group or assignment by manipulating IDs (mismatched IDs return 404).
- When a user is added to an organization-owned group, they are also added to the organization as a member, so group membership can't outrun organization membership. Existing organization members keep their role, and org admins can promote them at the organization level if needed.

### Screenshots of the UI changes (If any) -

---

## Code Understanding and AI Usage

**Did you use AI generated code (ChatGPT, Claude, Copilot, etc.) in any part of this PR? **
- [x] No, I wrote all the code myself
- [ ] Yes, I used AI assistance (continue below)

**If you used AI assistance:**
- [ ] I have reviewed every single line of the AI-generated code
- [ ] I can explain the purpose and logic of each function/component I added
- [ ] I have tested edge cases and understand how the code handles them
- [ ] I have modified the AI output to follow this project's coding standards and conventions

**Explain your implementation approach:**
Problem: Organization-owned groups use organization-scoped URLs, but their assignments still used flat routes, losing the organization context. Only a group's primary mentor could manage its assignments, so organization admins and assigned mentors couldn't manage assignments other mentors created. Users could also be added to an organization's group without belonging to the organization.

Approach: I nested assignments (plus reopen/close/start) under the organization-scoped groups so they get organization-scoped URLs, while standalone groups keep their flat URLs. I made AssignmentPolicy organization-aware with a private org_manage? method that returns false for non-organization groups and otherwise delegates to OrganizationGroupPolicy#manage_assignments? (organization admins, assigned mentors, and site admins), so anyone who runs an organization group's coursework can manage its assignments, while group settings (renaming, adding members/mentors) stay restricted to organization admins and the primary mentor via the stricter manage?. Controller redirects and view links use shared helpers in the OrganizationScopedRedirect concern that pick the scoped or flat path based on the group. Mailers do the same but check the feature flag globally, since there's no current user in that context. For the membership gap, adding a user to an organization-owned group now also adds them to the organization as a member via find_or_create_by, mirroring the existing group-invite flow so existing members aren't downgraded.

---

## Checklist before requesting a review
- [x] I have added proper PR title and linked to the issue
- [x] I have performed a self-review of my code
- [x] **I can explain the purpose of every function, class, and logic block I added**
- [x] I understand why my changes work and have tested them thoroughly
- [x] I have considered potential edge cases and how my code handles them
- [x] If it is a core feature, I have added thorough tests
- [x] My code follows the project's style guidelines and conventions

---

Note: Please check **Allow edits from maintainers** if you would like us to assist in the PR.

## Summary by CodeRabbit

* **New Features**
  * Added full organization-scoped group management, including viewing, editing, deleting, invitations, and assignments.
  * Added organization-aware links and redirects across group, assignment, and email workflows.
  * Redesigned organization group listings with member counts and management actions.
  * Added a back-to-organization link for organization-owned groups.

* **Bug Fixes**
  * Restricted organization overviews and group access based on membership and permissions.
  * Improved authorization for administrators, mentors, and members.
  * Prevented access through mismatched organization URLs.
  * Ensured invited group members receive appropriate organization membership.

### CODE REVIEW

File: app/controllers/groups_controller.rb [hunk 1]
```diff
@@ -126,6 +131,28 @@ def check_edit_access
       end
     end
 
+    def verify_organization_scope
+      return if params[:organization_id].blank?
+      return if @group.organization&.to_param == params[:organization_id]
+
+      redirect_to group_path(@group)
+    end
```
[the remark this answers is not in the document]
Comment (naman79820) [reply app/controllers/groups_controller.rb]: Authorization is already enforced through check_show_access and the redirect intentionally sends org-scoped mismatches to the canonical group path @coderabbitai

[the remark this answers is not in the document]
Comment (naman79820) [reply app/controllers/groups_controller.rb]: if i make my def set_group
  @group =
    if params[:organization_id].present?
      Organization.friendly.find(params.expect(:organization_id))
                  .groups.find(params.expect(:id))
    else
      Group.find(params.expect(:id))
    end
end
then i can remove my whole before action verify_organization_scope with its private method right? as it will become useless and dead code anyway @coderabbitai

File: config/routes.rb [hunk 2]
```diff
@@ -36,7 +36,19 @@
     end
     resources :organization_members, only: %i[create update destroy]
     delete :leave, to: "organization_members#leave"
-    resources :groups, only: %i[new create]
+    resources :groups, only: %i[index new create show edit update destroy] do
+      resources :assignments, except: %i[index] do
+        member do
+          get :reopen
+          put :close
+          get :start
```
[the remark this answers is not in the document]
Comment (naman79820) [reply config/routes.rb:44]: Actually i didn't introduce these routes, it already exist in assignment routes, i only mirrored it @coderabbitai

**Review by vedant-jain03 — APPROVED** [review]

**Comment by vedant-jain03** [comment]
Make the CI happy, we can merge then!

**Comment by naman79820** [comment]
> Make the CI happy, we can merge then!

done @vedant-jain03
